mssql_base.py

```diff
@@ -1,91 +1,105 @@
-        super().__init__()
-        self.cnxn = None
-        self.cursor = None
-        self.isConnecting = False
-        # maintains a LUT index by table containing a list of columns for it
-        self._table_colums = {}
-
-    def configure(self, context):
-        super().configure(context)
-        self.connect()
-
-    def connect(self):
-        self.isConnecting = True
-        cnxn_string = (
-            'DRIVER={};'
-            'PORT={};'
-            'SERVER={};'
-            'DATABASE={};'
-            'UID={};'
-            'MARS_Connection={};'
-            'PWD={}').format(
-                '{ODBC Driver 17 for SQL Server}',
-                self.port(),
-                self.server(),
-                self.database(),
-                self.credentials().userid(),
-                'yes' if self.mars() else 'no',
-                self.credentials().password())
-        self.logger.debug('Connecting: {}'.format(cnxn_string))
-        self.cnxn = pyodbc.connect(cnxn_string)
-        self.isConnecting = False
-
-    def disconnect(self):
-        if self.cnxn:
-            self.cnxn.close()
-            self.cnxn = None
-
-    def stop(self):
-        super().stop()
-        self.disconnect()
-
-    def validate_column(self, column, table, cursor):
-        """ Makes sure column belongs to table
-
-        Args:
-            column (str): column in question
-            table (str): table name
-            cursor (pyodbc.Cursor): active cursor
-
-        Returns:
-            True/False
-        """
-        columns = self._table_colums.get(table)
-        if columns is None:
-            columns = [column.column_name for column in cursor.columns(table)]
-            self._table_colums[table] = columns
-
-        if column not in columns:
-            cursor.close()
-            raise ValueError(
-                '\"{}\" is not a valid column in table \"{}\".'.format(
-                    column, table))
-
-        return column
+import pyodbc
+from nio.block.base import Block
+from nio.properties import (StringProperty, IntProperty, BoolProperty, ObjectProperty, PropertyHolder)
+from nio.util.discovery import not_discoverable
+
+
+class Connection(PropertyHolder):
+    server = StringProperty(title='Server', default='[[MSSQL_SERVER]]', order=1)
+    port = IntProperty(title='Port', default='[[MSSQL_PORT]]', order=2)
+    database = StringProperty(title='Database', default='[[MSSQL_DB]]', order=3)
+    user_id = StringProperty(title='User ID', allow_none=True, default='[[MSSQL_USER]]', order=4)
+    password = StringProperty(title='Password', allow_none=True, default='[[MSSQL_PWD]]', order=5)
+
+class Mars(PropertyHolder):
+    enabled = BoolProperty(title='Enable MARS?', default=True, order=1)
+
+@not_discoverable
+class MSSQLBase(Block):
+    connection = ObjectProperty(Connection, title='Database Connection', order=50)
+    mars = ObjectProperty(Mars, title='Multiple Active Result Sets (MARS)', order=60)

```

Comment:
I don't think this needs to be an object property (it only has one object) and it should be advanced. Think this can be replaced with
```python
mars = BoolProperty(title='Enable MARS?', default=True, advanced=True)
```

Reply:
I put that in there because it looks nicer when it's wrapped in a fieldset... and I didn't hide it because it's a pretty major settin, at least for the RawQuery... so then I didn't want to hide it for some blocks, but not others...

Reply:
Haha well if it's not advanced I guess we could have just left it in the connection properties then. Seems like an advanced property to me but I'll take your word for it since you have used the blocks more 👍